woocommerce/woocommerce-ios
# [Order List] Fix iphone crash when Orders Split View flag is enabled

Closes: #11649 


## Description


The app would crash when the `splitViewInOrdersTab` was enabled, and the orders were used in a horizontally compact view (i.e. on iPhone, or in some split view/slide over presentations on iPad.)

The original crash was trivial to reproduce:

> 1. Go to Orders
> 2. Tap on any order from the list
> 3. Crash
– #10815 

The crash was non-trivial to trace, and you can see from the commit history of this PR that there were a few parts to it.

On iOS 16, it was sufficient to update the various split view function calls to use the modern column-based API, which is not enforced by Apple: f15eb12

However, when built from Xcode 15 on Xcode 17, the crash returned. In this case, it was due to our `UISplitViewController` being wrapped in a `UINavigationController`, as all our root tabs were. Nav controllers live inside split view controllers, and not (ideally) the other way around.

To resolve this without substantially changing how the `MainTabViewController` works, I've added a `WooTabContainerController`, which can be used to wrap view controllers at the root of a tab which should not be presented in a nav controller.

### Other changes
- We only auto-select the first order when the splitview is uncollapsed. Without this, we would sometimes see the phone skip over the Order List and go straight to the Order Details.
- The Order List is always responsible for auto-selection (not the OrdersRoot or OrdersSplitViewWrapper.) 

## Testing instructions


The main thing with this PR is to test on a wide array of devices and with the feature flag `splitViewInOrdersTab` both true and false. 

When the flag's set to false, _everything_ should work. When it's true, there are some known issues with the split view, which are not relevant to this PR – the main thing is that it shouldn't crash!

Devices:
iPad on iOS 15
iPad on iOS 16
iPad on iOS 17 
(include OS-based split view multitasking for smaller size classes)

iPhone on iOS 15
iPhone on iOS 16
iPhone on iOS 17
(if possible, test Plus/Pro Max phone sizes in landscape for horizontally regular size classes)

Here are some scenarios to cover:

### Navigating to an order

1. Launch the app
2. Tap `Orders`
3. Verify that you can view the order details
4. Tap `<` (if available)
5. Select another order
6. Verify that you can view the second order

### Navigating to an order after switching store

1. Launch the app
2. Tap `Orders`
3. Verify that you can view the order details
4. Switch to another store
5. Tap `Orders`, and navigate to an order
6. Verify that you can view the second store's orders

## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/2472348/be449c2a-cead-40f8-9b36-d23fbfa18cff



https://github.com/woocommerce/woocommerce-ios/assets/2472348/32000e8b-abe0-4d52-8ca4-592beb515cee





---
